omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java

```diff
@@ -143,45 +149,78 @@ protected ModelAndView onSubmit(HttpServletRequest request, HttpServletResponse
 					httpSession.setAttribute(WebConstants.OPENMRS_MSG_ATTR, "auth.question.fill");
-					//request.setAttribute("secretQuestion", user.getSecretQuestion());
+					request.setAttribute("secretQuestion", getRandomFakeSecretQuestion(username));
+				} else {
+					String secretQuestion = Context.getUserService().getSecretQuestion(user);
+					if (secretQuestion == null || secretQuestion.equals("")) {
+						httpSession.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, "auth.question.empty");
+					} else if (secretQuestion != null && Context.getUserService().isSecretAnswer(user, secretAnswer)) {
+						
+						String randomPassword = getRandomPassword();
+						
+						try {
+							Context.addProxyPrivilege(PrivilegeConstants.EDIT_USER_PASSWORDS);
+							Context.getUserService().changePassword(user, randomPassword);
+						} finally {
+							Context.removeProxyPrivilege(PrivilegeConstants.EDIT_USER_PASSWORDS);
+						}
+						
+						httpSession.setAttribute("resetPassword", randomPassword);
+						httpSession.setAttribute(WebConstants.OPENMRS_MSG_ATTR, "auth.password.reset");
+						Context.authenticate(username, randomPassword);
+						httpSession.setAttribute("loginAttempts", 0);
+						
+						return new ModelAndView(new RedirectView(request.getContextPath() + "/options.form#Change Login Info"));
+					} else {
+						httpSession.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, "auth.answer.invalid");
+						httpSession.setAttribute(WebConstants.OPENMRS_MSG_ATTR, "auth.question.fill");
+						request.setAttribute("secretQuestion", secretQuestion);
+					}
 				}
 			}
 		}
 		
 		loginAttemptsByIP.put(ipAddress, forgotPasswordAttempts);
+		
 		request.setAttribute("uname", username);
+		
 		return showForm(request, response, errors);
 	}
 	
-}
+	public String getRandomPassword() {
+		//Password should be of length at-least 8, must have 1 upper case letter and 1 number
+		return RandomStringUtils.randomAlphabetic(1).toUpperCase() + RandomStringUtils.randomAlphanumeric(8)+ RandomStringUtils.randomNumeric(1);
+	}
+	
+	public String getRandomFakeSecretQuestion(String username) {
+		
+		List<String> questions = new ArrayList<String>();
+		
+		questions.add(Context.getMessageSourceService().getMessage("What is your best friend's name?"));
+		questions.add(Context.getMessageSourceService().getMessage("What is your grandfather's home town?"));
+		questions.add(Context.getMessageSourceService().getMessage("What is your mother's maiden name?"));
+		questions.add(Context.getMessageSourceService().getMessage("What is your favorite band?"));
+		questions.add(Context.getMessageSourceService().getMessage("What is your first pet's name?"));
+		questions.add(Context.getMessageSourceService().getMessage("What is your brother's middle name?"));
+		questions.add(Context.getMessageSourceService().getMessage("Which city were you born in?"));
+		
+		int hashValueForName = username.hashCode();
```

Comment:
Apache's RandomUtils can easily give you a random number with a single method call